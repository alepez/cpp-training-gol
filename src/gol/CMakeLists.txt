add_library(gol STATIC)

target_include_directories(gol
        PUBLIC ${CMAKE_SOURCE_DIR}/src
)

target_link_libraries(gol
        PUBLIC spdlog::spdlog
        PUBLIC fmt::fmt
)

# Enable all warnings and treat them as errors, but only for the library
target_compile_options(gol
        PRIVATE -Wall
        PRIVATE -Wpedantic
        PRIVATE -Wextra
        PRIVATE -Werror
        PRIVATE -Wno-unused-parameter
)

add_executable(gol_test)

target_include_directories(gol_test
        PRIVATE ${CMAKE_SOURCE_DIR}/src
)

target_link_libraries(gol_test
        PRIVATE Catch2::Catch2WithMain
        PRIVATE ApprovalTests::ApprovalTests
        PRIVATE gol
)

function(add_module name)
    target_sources(gol PRIVATE ${name}.cpp)
    target_sources(gol_test PRIVATE ${name}_test.cpp)
endfunction()

add_module(foo)
add_module(bar)
add_module(gol)
