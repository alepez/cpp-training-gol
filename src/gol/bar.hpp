#pragma once

#include <SDL2/SDL.h>

namespace gol {

class Grid;

class SdlCanvas {
  public:
    SdlCanvas(int w, int h, int scale);

    ~SdlCanvas();

    bool draw(Grid& grid);

  private:
    SDL_Window* win;
    SDL_Event event;
};

}  // namespace gol