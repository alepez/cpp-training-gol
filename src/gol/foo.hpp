#pragma once

namespace gol {

class Grid;

class TerminalCanvas {
  public:
    TerminalCanvas(int w, int h, int scale);

    bool draw(Grid& grid);

  private:
    int w;
    int h;
    int scale;
};

}  // namespace gol