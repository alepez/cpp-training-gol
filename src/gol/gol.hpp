#pragma once

#include <vector>

namespace gol {

class Grid {
  public:
    Grid(int w, int h);

    int* data();

    int& at(int x, int y);

    int width() const {
        return w_;
    }

    int height() const {
        return h_;
    }

  private:
    int w_;
    int h_;
    std::vector<int> data_;
};

class Game {
  public:
    void update(Grid& grid);
};

}  // namespace gol
