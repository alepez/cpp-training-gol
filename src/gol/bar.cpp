#include "bar.hpp"

#include <SDL2/SDL.h>

#include <exception>

#include "gol.hpp"

namespace gol {

SdlCanvas::SdlCanvas(int w, int h, int scale) {
    if (SDL_Init(SDL_INIT_VIDEO) != 0) {
        std::terminate();
    }

    win = SDL_CreateWindow(
        "Conway's Game of Life",
        0,
        0,
        scale * w,
        scale * h,
        4
    );

    if (!win) {
        std::terminate();
    }
}

bool SdlCanvas::draw(Grid& grid) {
    int x;
    int y;
    bool running = true;
    for (x = 0; x < SDL_GetWindowSurface(win)->w; ++x)
        for (y = 0; y < SDL_GetWindowSurface(win)->h; ++y)
            ((int*)(SDL_GetWindowSurface(win)->pixels)
            )[x + y * SDL_GetWindowSurface(win)->w] = 0x00FFFFFF
                * grid.at(
                    x / (SDL_GetWindowSurface(win)->w / grid.width()),
                    y / (SDL_GetWindowSurface(win)->h / grid.height())
                );
    SDL_UpdateWindowSurface(win);
    while (SDL_PollEvent(&event))
        if (event.type == SDL_QUIT)
            running = false;
    SDL_Delay(40);
    return running;
}

SdlCanvas::~SdlCanvas() {
    SDL_DestroyWindow(win);
    SDL_Quit();
}

}  // namespace gol