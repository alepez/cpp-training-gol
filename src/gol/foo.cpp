#include "foo.hpp"

#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <thread>

#include "gol.hpp"

using namespace std::chrono_literals;

namespace gol {

TerminalCanvas::TerminalCanvas(int w, int h, int scale) :
    w(w), h(h), scale(scale) {
#ifdef _WIN32
    system("cls");
    system("color 0F");
#else
    system("clear");
#endif
}

bool TerminalCanvas::draw(Grid& grid) {
    int x;
    int y;
    for (y = 0; y < h; ++y) {
        for (x = 0; x < w; ++x)
            printf(grid.at(x, y) ? "X" : " ");
        printf("\n");
    }
    std::this_thread::sleep_for(40ms);
#ifdef _WIN32
    system("cls");
#else
    system("clear");
#endif
    return true;
}

}  // namespace gol