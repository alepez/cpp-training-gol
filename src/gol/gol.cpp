#include "gol.hpp"

namespace gol {

#define cell(g, x, y) g[((h + y) % h) * w + ((w + x) % w)]

void update_grid(int* grid, int* grid_tmp, int w, int h) {
    int x;
    int y;
    int adjacent_count;
    for (x = 0; x < w; ++x) {
        for (y = 0; y < h; ++y) {
            adjacent_count = cell(grid, x - 1, y - 1) + cell(grid, x - 1, y + 0)
                + cell(grid, x - 1, y + 1) + cell(grid, x + 0, y - 1)
                + cell(grid, x + 0, y + 1) + cell(grid, x + 1, y - 1)
                + cell(grid, x + 1, y + 0) + cell(grid, x + 1, y + 1);
            cell(grid_tmp, x, y) = adjacent_count == 3
                || (cell(grid, x, y) && adjacent_count == 2);
        }
    }
    for (x = 0; x < w * h; ++x) {
        grid[x] = grid_tmp[x];
    }
}

double something_really_obscure(double x) {
    return x * x;
}

}  // namespace gol