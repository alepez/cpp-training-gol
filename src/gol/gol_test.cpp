#include "Approvals.h"
#define APPROVALS_CATCH2_V3
#include <ApprovalTests.hpp>
#include <gol/gol.hpp>
#include <string>

using ApprovalTests::Approvals;
using gol::something_really_obscure;
using std::back_inserter;
using std::string;
using std::to_string;
using std::transform;
using std::vector;
