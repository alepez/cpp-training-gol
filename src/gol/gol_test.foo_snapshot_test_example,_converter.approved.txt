This is the header


[4] = 16
[8] = 64
[15] = 225
[16] = 256
[23] = 529
[42] = 1764

